File: abstractions/typescript/src/requestInfo.ts
import { HttpMethod } from "./httpMethod";
import { ReadableStream } from 'web-streams-polyfill/es2018';
import { Parsable } from "./serialization";
import { HttpCore } from "./httpCore";
import { MiddlewareOption } from "./middlewareOption";

/** This class represents an abstract HTTP request. */
export class RequestInfo {
    /** The URI of the request. */
    public URI?: string;
    public setUri(currentPath: string, pathSegment: string, isRawUrl: boolean) : void {
        if(isRawUrl) {
            const questionMarkSplat = currentPath.split('?');
            const schemeHostAndPath = questionMarkSplat[0];
            this.URI = schemeHostAndPath;
            if(questionMarkSplat.length > 1) {
                const queryString = questionMarkSplat[1];
                queryString?.split('&').forEach(queryPair => {
                    const keyValue = queryPair.split('=');
                    if(keyValue.length > 1) {
                        const key = keyValue[0];
                        if(key) {
                            this.queryParameters.set(key, keyValue[1]);
                        }
                    }
                });
            }
        } else {
            this.URI = currentPath + pathSegment;
        }
    }
    /** The HTTP method for the request */
    public httpMethod?: HttpMethod;
    /** The Request Body. */
    public content?: ReadableStream;
    /** The Query Parameters of the request. */
    public queryParameters: Map<string, string | number | boolean | undefined> = new Map<string, string | number | boolean | undefined>(); //TODO: case insensitive
    /** The Request Headers. */
    public headers: Map<string, string> = new Map<string, string>(); //TODO: case insensitive
    private _middlewareOptions = new Map<string, MiddlewareOption>(); //TODO: case insensitive
    /** Gets the middleware options for the request. */
    public getMiddlewareOptions() { return this._middlewareOptions.values(); }
    public addMiddlewareOptions(...options: MiddlewareOption[]) {
        if(!options || options.length === 0) return;
        options.forEach(option => {
            this._middlewareOptions.set(option.getKey(), option);
        });
    }
    /** Removes the middleware options for the request. */
    public removeMiddlewareOptions(...options: MiddlewareOption[]) {
        if(!options || options.length === 0) return;
        options.forEach(option => {
            this._middlewareOptions.delete(option.getKey());
        });
    }
    private static binaryContentType = "application/octet-stream";
    private static contentTypeHeader = "Content-Type";
    /**
     * Sets the request body from a model with the specified content type.
     * @param values the models.
     * @param contentType the content type.
     * @param httpCore The core service to get the serialization writer from.
     * @typeParam T the model type.
     */
    public setContentFromParsable = <T extends Parsable>(httpCore?: HttpCore | undefined, contentType?: string | undefined, ...values: T[]): void => {
        if(!httpCore) throw new Error("httpCore cannot be undefined");
        if(!contentType) throw new Error("contentType cannot be undefined");
        if(!values || values.length === 0) throw new Error("values cannot be undefined or empty");

        const writer = httpCore.getSerializationWriterFactory().getSerializationWriter(contentType);
        this.headers.set(RequestInfo.contentTypeHeader, contentType);
        if(values.length === 1) 
            writer.writeObjectValue(undefined, values[0]);
        else
            writer.writeCollectionOfObjectValues(undefined, values);
        this.content = writer.getSerializedContent();
    }
    /**
     * Sets the request body to be a binary stream.
     * @param value the binary stream
     */
    public setStreamContent = (value: ReadableStream): void => {
        this.headers.set(RequestInfo.contentTypeHeader, RequestInfo.binaryContentType);
        this.content = value;
    }
    /**
     * Sets the request headers from a raw object.
     * @param headers the headers.
     */
    public setHeadersFromRawObject = (h: object) : void => {
        Object.entries(h).forEach(([k, v]) => {
            this.headers.set(k, v as string);
        });
    }
    /**
     * Sets the query string parameters from a raw object.
     * @param parameters the parameters.
     */
    public setQueryStringParametersFromRawObject = (q: object): void => {
        Object.entries(q).forEach(([k, v]) => {
            this.headers.set(k, v as string);
        });
    }
}

File: abstractions/dotnet/src/RequestInfo.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Kiota.Abstractions.Serialization;

namespace Microsoft.Kiota.Abstractions
{
    /// <summary>
    ///     This class represents an abstract HTTP request.
    /// </summary>
    public class RequestInfo
    {
        /// <summary>
        ///  The URI of the request.
        /// </summary>
        public Uri URI { get; set; }
        /// <summary>
        /// Sets the URI of the request.
        /// </summary>
        /// <param name="currentPath">the current path (scheme, host, port, path, query parameters) of the request.</param>
        /// <param name="pathSegment">the segment to append to the current path.</param>
        /// <param name="isRawUrl">whether the path segment is a raw url. When true, the segment is not happened and the current path is parsed for query parameters.</param>
        /// <exception cref="UriFormatException">Thrown when the built URI is an invalid format.</exception>
        public void SetURI(string currentPath, string pathSegment, bool isRawUrl)
        {
            if (isRawUrl)
            {
                if(string.IsNullOrEmpty(currentPath))
                    throw new ArgumentNullException(nameof(currentPath));
                var parseUri = new Uri(currentPath);
                var parseQueryString = parseUri.Query.TrimStart('?'); //remove leading ? if needed
                foreach(var qsp in parseQueryString.Split('&').Select(x => x.Split('=')).Where(x => !string.IsNullOrEmpty(x[0]))) {
                    QueryParameters.Add(qsp[0], qsp.Length > 1 ? qsp[1] : null);
                }
                URI = new Uri(parseUri.GetComponents(UriComponents.SchemeAndServer | UriComponents.Path, UriFormat.Unescaped));
            }
            else
            {
                URI = new Uri(currentPath + pathSegment);
            }
        }
        /// <summary>
        ///  The <see cref="HttpMethod">HTTP method</see> of the request.
        /// </summary>
        public HttpMethod HttpMethod { get; set; }
        /// <summary>
        /// The Query Parameters of the request.
        /// </summary>
        public IDictionary<string, object> QueryParameters { get; set; } = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
        /// <summary>
        /// The Request Headers.
        /// </summary>
        public IDictionary<string, string> Headers { get; set; } = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        /// <summary>
        /// The Request Body.
        /// </summary>
        public Stream Content { get; set; }
        private Dictionary<string, IMiddlewareOption> _middlewareOptions = new Dictionary<string, IMiddlewareOption>(StringComparer.OrdinalIgnoreCase);
        /// <summary>
        /// Gets the middleware options for this request. Options are unique by type. If an option of the same type is added twice, the last one wins.
        /// </summary>
        public IEnumerable<IMiddlewareOption> MiddlewareOptions { get { return _middlewareOptions.Values; } }
        /// <summary>
        /// Adds a middleware option to the request.
        /// </summary>
        /// <param name="options">The middleware option to add.</param>
        public void AddMiddlewareOptions(params IMiddlewareOption[] options)
        {
            if(!(options?.Any() ?? false)) return; // it's a no-op if there are no options and this avoid having to check in the code gen.
            foreach(var option in options.Where(x => x != null))
                if(!_middlewareOptions.TryAdd(option.GetType().FullName, option))
                    _middlewareOptions[option.GetType().FullName] = option;
        }
        /// <summary>
        /// Removes given middleware options from the current request.
        /// </summary>
        /// <param name="options">Middleware options to remove.</param>
        public void RemoveMiddlewareOptions(params IMiddlewareOption[] options)
        {
            if(!options?.Any() ?? false) throw new ArgumentNullException(nameof(options));
            foreach(var optionName in options.Where(x => x != null).Select(x => x.GetType().FullName))
                _middlewareOptions.Remove(optionName);
        }
        private const string BinaryContentType = "application/octet-stream";
        private const string ContentTypeHeader = "Content-Type";
        /// <summary>
        /// Sets the request body to a binary stream.
        /// </summary>
        /// <param name="content">The binary stream to set as a body.</param>
        public void SetStreamContent(Stream content)
        {
            Content = content;
            Headers.Add(ContentTypeHeader, BinaryContentType);
        }
        /// <summary>
        /// Sets the request body from a model with the specified content type.
        /// </summary>
        /// <param name="coreService">The core service to get the serialization writer from.</param>
        /// <param name="items">The models to serialize.</param>
        /// <param name="contentType">The content type to set.</param>
        /// <typeparam name="T">The model type to serialize.</typeparam>
        public void SetContentFromParsable<T>(IHttpCore coreService, string contentType, params T[] items) where T : IParsable
        {
            if(string.IsNullOrEmpty(contentType)) throw new ArgumentNullException(nameof(contentType));
            if(coreService == null) throw new ArgumentNullException(nameof(coreService));
            if(items == null || !items.Any()) throw new InvalidOperationException($"{nameof(items)} cannot be null or empty");

            using var writer = coreService.SerializationWriterFactory.GetSerializationWriter(contentType);
            if(items.Count() == 1)
                writer.WriteObjectValue(null, items[0]);
            else
                writer.WriteCollectionOfObjectValues(null, items);
            Headers.Add(ContentTypeHeader, contentType);
            Content = writer.GetSerializedContent();
        }
    }
}
